canonical-web-and-design/ubuntu.com
# Optimise Dockerfile using multistage builds

I've done a couple of things:

1. We will no longer need to run `./run build` before `docker build .`.
   Simply running `docker build .` will build everything needed.
2. The build is optimised in the following ways
    - Yarn and Python dependencies are installed in independent build steps, meaning they will be cached if they haven't changed
    - We make use of experimental buildkit caching so that both pip and yarn can reuse previously downloaded packages
    - The "build-css" and "build-js" stages are also separate steps, so similarly these won't need to be re-run if they haven't changed
    - The experimental buildkit will also run any parts of the build that can be run concurrently - so e.g. yarn dependencies will download at the same time as python dependencies.
   This should speed up subsequent builds by quite a bit.

QA
--

First, let me show you what a problem this was on `master`:

``` bash
$ git checkout master  # master branch
$ ./run clean && ./run clean-cache && docker system prune -f  # clean everything our
$ time sh -c './run build && docker build --build-arg TALISKER_REVISION_ID=sdf --tag ubuntu .'  # Build once
...
real    2m30.541s
user    0m1.741s
sys     0m0.990s
$ echo '' >> templates/base_index.html  # Make a change
$ time sh -c './run build && docker build --build-arg TALISKER_REVISION_ID=sdf --tag ubuntu .'  # Build again
...
real    1m3.761s
user    0m1.176s
sys     0m0.723s
```

The second time is still a long time.

Now try with the new branch:

``` bash
$ git checkout .  # Undo the change to the template
$ git checkout optimise-dockerfile  # Switch to this branch
$ docker system prune -f; docker builder prune -f  # Clean things out
$ DOCKER_BUILDKIT=1 time docker build  --tag ubuntu . --no-cache  # Build without cache
...
.. 1:22.19elapsed ..
$ echo '' >> templates/base_index.html  # Make a change
$ DOCKER_BUILDKIT=1 time docker build  --tag ubuntu .  # Build again
.. 0:03.83elapsed ..
```

Both the initial and second build should be a lot quicker.

Now check the site still runs and looks cool:

``` bash
docker run -ti -p 8001:80 ubuntu
```

Go to http://localhost:8001.
